src/compiler/checker.ts

```diff
@@ -16773,7 +16773,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
             node = parent;
         }
-        return constraints ? getSubstitutionType(type, getIntersectionType(constraints)) : type;
+        if (!constraints) {
+            return type;
+        }
+        if (isNoInferType(type)) {
+            constraints = append(constraints, (type as SubstitutionType).baseType);
+            return getNoInferType(getIntersectionType(constraints));
+        }

```

Comment:
Alternatively, I was thinking about handling the substitution types created here today (with `NoInfer` base types and non-unknown constraints) in `isNoInferType` (or at least handling such at the top of `inferFromTypes`). I concluded this is better and less error-prone though